File: src/main/java/com/baulsupp/oksocial/network/InterfaceSocketFactory.java
package com.baulsupp.oksocial.network;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.NetworkInterface;
import java.net.Socket;
import java.net.SocketException;
import java.net.UnknownHostException;
import java.util.Optional;
import javax.net.SocketFactory;

import static java.util.Optional.empty;
import static java.util.Optional.of;

public class InterfaceSocketFactory extends SocketFactory {
  private InetAddress localAddress;
  private SocketFactory systemFactory = SocketFactory.getDefault();

  public InterfaceSocketFactory(InetAddress localAddress) {
    this.localAddress = localAddress;
  }

  @Override public Socket createSocket() throws IOException {
    Socket s = systemFactory.createSocket();
    s.bind(new InetSocketAddress(localAddress, 0));
    return s;
  }

  @Override public Socket createSocket(String host, int port) throws IOException {
    return systemFactory.createSocket(host, port, localAddress, 0);
  }

  @Override public Socket createSocket(InetAddress address, int port) throws IOException {
    return systemFactory.createSocket(address, port, localAddress, 0);
  }

  @Override public Socket createSocket(String host, int port, InetAddress localAddr, int localPort)
      throws IOException {
    return systemFactory.createSocket(host, port, localAddr, localPort);
  }

  @Override
  public Socket createSocket(InetAddress address, int port, InetAddress localAddr, int localPort)
      throws IOException {
    return systemFactory.createSocket(address, port, localAddr, localPort);
  }

  public static Optional<SocketFactory> byName(String ipOrInterface) throws SocketException {
    InetAddress localAddress;
    try {
      // example 192.168.0.51
      localAddress = InetAddress.getByName(ipOrInterface);
    } catch (UnknownHostException uhe) {
      // example en0
      NetworkInterface networkInterface = NetworkInterface.getByName(ipOrInterface);

      if (networkInterface == null) {
        return empty();
      }

      localAddress = networkInterface.getInetAddresses().nextElement();
    }

    return of(new InterfaceSocketFactory(localAddress));
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/network/InterfaceSocketFactory.kt
package com.baulsupp.oksocial.network

import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.NetworkInterface
import java.net.Socket
import java.net.SocketException
import java.net.UnknownHostException
import java.util.Optional
import java.util.Optional.empty
import java.util.Optional.of
import javax.net.SocketFactory

class InterfaceSocketFactory(private val localAddress: InetAddress) : SocketFactory() {
    private val systemFactory = SocketFactory.getDefault()

    @Throws(IOException::class)
    override fun createSocket(): Socket {
        val s = systemFactory.createSocket()
        s.bind(InetSocketAddress(localAddress, 0))
        return s
    }

    @Throws(IOException::class)
    override fun createSocket(host: String, port: Int): Socket {
        return systemFactory.createSocket(host, port, localAddress, 0)
    }

    @Throws(IOException::class)
    override fun createSocket(address: InetAddress, port: Int): Socket {
        return systemFactory.createSocket(address, port, localAddress, 0)
    }

    @Throws(IOException::class)
    override fun createSocket(host: String, port: Int, localAddr: InetAddress,
                              localPort: Int): Socket {
        return systemFactory.createSocket(host, port, localAddr, localPort)
    }

    @Throws(IOException::class)
    override fun createSocket(address: InetAddress, port: Int, localAddr: InetAddress,
                              localPort: Int): Socket {
        return systemFactory.createSocket(address, port, localAddr, localPort)
    }

    companion object {

        @Throws(SocketException::class)
        fun byName(ipOrInterface: String): Optional<SocketFactory> {
            val localAddress = try {
                // example 192.168.0.51
                InetAddress.getByName(ipOrInterface)
            } catch (uhe: UnknownHostException) {
                // example en0
                val networkInterface = NetworkInterface.getByName(ipOrInterface) ?: return empty()

                networkInterface.inetAddresses.nextElement()
            }

            return of(InterfaceSocketFactory(localAddress))
        }
    }
}
